File: Data Structures & Algorithms/balanced-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# from collectios import deque

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        return self.height(root) != -1
            

    def height(self,node: Optional[TreeNode]) -> int:
        if not node:
            return 0
        
        left = self.height(node.left)
        if left == -1:
            return -1

        right = self.height(node.right)
        if right == -1:
            return -1

        if abs(left-right) > 1:
            return -1
        
        return 1 + max(right,left)

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int maxPathSum(TreeNode* root) {
        int max_path = INT_MIN;
        dfs(root,max_path);
        return max_path;
    }

    int dfs(TreeNode* node, int& max_path){
        if(!node) return 0;

        int left = dfs(node->left,max_path);
        int right = dfs(node->right,max_path);

        max_path = max(max_path,node->val + max(0,left) + max(0,right));
        return node->val + max(0,max(left,right));

    }
};
